### ISSUE TITLE
Poisson, gamma, tweedie losses and offsets

### ISSUE THREAD

**Issue by thenomemac** [dmlc/xgboost#681]
It would be really useful to have poisson and other tweedie family loss functions integrated into xgboost with offsets. This is already implemented in R's GBM library and is widely used. Not sure if my C skills are complainant enough to tackle this. But I'm willing to try if pointed in the right directly. Currently working to add these features to sklearn myself.

**Comment by tqchen** [comment]
poission is already supported, there is a plugin system in to add new objectives and others which is a good start place https://github.com/dmlc/xgboost/tree/brick/plugin

**Comment by spedygiorgio** [comment]
I understant that plugins can help for this task, but most final user do not have the necessary skills to write that function. Its a real pity offset cannot be used for poisson distribution cause this makes xgboost not a choise for rate models (that are very frequent in actuarial and ecological modeling). An offset addition for poisson regression would really help.

**Comment by spedygiorgio** [comment]
for the benefit of all, according to [crossvalidated](http://stats.stackexchange.com/questions/223939/poisson-xgboost-with-exposure/237576#237576) a possible solution is `setinfo(xgbMatrix,"base_margin",log(Insurance$Holders)).` 
But I beg the package authors'  @tqchen  to provide some documentation for it cause this application is very relevant...

**Comment by shearerp** [comment]
If I can use Poisson log link with count:poisson, as I can do with tweedie, then the base_margin technique will work. @tqchen , is it possible to use log link in conjunction with Poisson objective?

**Comment by thenomemac** [comment]
I know for a fact from simulation this doesn't work. You are not correctly
weighting the count data in the loss functions when normalizing exposures
by dividing through. In poisson variance is proportional to exposure and
you're breaking this link the model will no converge to the same solution
as a GBM with offsets such as the R implement. Have to take the time to
workout the proof to demo this for this discussion thread one of these
days.

**Comment by shearerp** [comment]
@thenomemac I've looked at the formulas that describe xgboost's implementation of boosting, and I believe the base_margin approach is correct, but only if the log link is used with the Poisson objective. Hence my request to Tianqi.

**Comment by spedygiorgio** [comment]
I have used the log exposure as base margin for poisson on a quite big
dataset ( 10 mln records 200 features) where the base event rate was around
1%. On the 10% test set the predict method gave consistent results. I have
not tested tweedie regression

Il ven 28 apr 2017, 00:47 shearerp <notifications@github.com> ha scritto:

**Comment by thenomemac** [comment]
I'll check k it out in more detail could be different from when I looked
last or I'm wrong. Thanks.

**Comment by shearerp** [comment]
I dug into the code a bit and I think log link is the default. See line 184 of https://github.com/dmlc/xgboost/blob/d581a3d0e7fbc8e73dbeb6f81fdc1e101606737e/src/objective/regression_obj.cc

The gradient is set as
`(std::exp(p) - y) * w`
This is only correct for a Poisson likelihood if p is the linear predictor on the log-link scale.
